x-pack/plugins/enterprise_search/server/lib/indices/fetch_indices.ts

```diff
@@ -70,133 +40,63 @@ export const fetchIndices = async (
   // If we provide an override pattern with alwaysShowSearchPattern we get everything and filter out hiddens.
   const expandWildcards: ExpandWildcard[] =
     returnHiddenIndices || alwaysShowSearchPattern ? ['hidden', 'all'] : ['open'];
-  const totalIndices = await client.asCurrentUser.indices.get({
-    expand_wildcards: expandWildcards,
-    // for better performance only compute aliases and settings of indices but not mappings
-    features: ['aliases', 'settings'],
-    // only get specified index properties from ES to keep the response under 536MB
-    // node.js string length limit: https://github.com/nodejs/node/issues/33960
-    filter_path: ['*.aliases', '*.settings.index.hidden'],
-    index: indexPattern,
-  });
-
-  // Index names that with one of their aliases match with the alwaysShowSearchPattern
-  const alwaysShowPatternMatches = new Set<string>();
-
-  const indexAndAliasNames = Object.keys(totalIndices).reduce((accum, indexName) => {
-    accum.push(indexName);
-
-    if (includeAliases) {
-      const aliases = Object.keys(totalIndices[indexName].aliases!);
-      aliases.forEach((alias) => {
-        accum.push(alias);
-
-        // Add indexName to the set if an alias matches the pattern
-        if (alwaysShowSearchPattern && alias.startsWith(alwaysShowSearchPattern)) {
-          alwaysShowPatternMatches.add(indexName);
-        }
-      });
-    }
-    return accum;
-  }, [] as string[]);
-
-  const indicesNames = returnHiddenIndices
-    ? Object.keys(totalIndices)
-    : Object.keys(totalIndices).filter(
-        (indexName) => !(totalIndices[indexName]?.settings?.index?.hidden === 'true')
-      );
+
+  const { allIndexMatches, indexAndAliasNames, indicesNames, alwaysShowMatchNames } =
+    await getIndexData(
+      client,
+      indexPattern,
+      expandWildcards,
+      returnHiddenIndices,
+      includeAliases,
+      alwaysShowSearchPattern
+    );
+
   if (indicesNames.length === 0) {
     return [];
   }
 
-  const { indices: indicesStats = {} } = await client.asCurrentUser.indices.stats({
-    expand_wildcards: expandWildcards,
-    index: indexPattern,
-    metric: ['docs', 'store'],
-  });
-
-  // TODO: make multiple batched requests if indicesNames.length > SOMETHING
-  const { index: indexPrivileges } = await client.asCurrentUser.security.hasPrivileges({
-    index: [
-      {
-        names: indexAndAliasNames,
-        privileges: ['read', 'manage'],
-      },
-    ],
-  });
+  const indicesStats = await fetchIndexStats(client, indexPattern, expandWildcards);
 
-  const indexCounts = await fetchIndexCounts(client, indexAndAliasNames);
+  const indexPrivileges = await fetchIndexPrivileges(client, indexAndAliasNames);
```

Comment:
Is it worth doing any of these fetch calls in parallel? seems like this `fetchIndexPrivileges` & `fetchIndexCounts` could be made at the same time?

```typescript
const [indexPrivileges, indexCounts] = await Promise.all([
  fetchIndexPrivileges(client, indexAndAliasNames),
  fetchIndexCounts(client, indexAndAliasNames),
]);
```